Should we use panic or return error?

## Question
<!-- You can ask any question about this project -->
Now, we use panic and recover to handle the error. However, `return error` can alse do that. This is a classic question in the `go` language. `Should we use panic or return error?`

There is a discussion [here ](https://stackoverflow.com/questions/44504354/should-i-use-panic-or-return-error) which may help us.

And I agree with the following sentence:

> Generally speaking, in go, if you're not sure, return an error. Panics should be pretty rare.

Just  a case as follows:
```
# dfget-go --url "http://test" --output ./resource.png --node "127.0.0.1"
panic: invalid url: http://test [recovered]
	panic: (*logrus.Entry) (0x78bbc0,0xc42008c870)

goroutine 1 [running]:
github.com/alibaba/Dragonfly/vendor/github.com/sirupsen/logrus.Entry.log(0xc42008c690, 0xc420081dd0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/tmp/dragonfly/build/src/github.com/alibaba/Dragonfly/vendor/github.com/sirupsen/logrus/entry.go:126 +0x318
github.com/alibaba/Dragonfly/vendor/github.com/sirupsen/logrus.(*Entry).Panic(0xc42008c820, 0xc420183d28, 0x1, 0x1)
	/tmp/dragonfly/build/src/github.com/alibaba/Dragonfly/vendor/github.com/sirupsen/logrus/entry.go:196 +0xaa
github.com/alibaba/Dragonfly/vendor/github.com/sirupsen/logrus.(*Logger).Panic(0xc42008c690, 0xc420055d28, 0x1, 0x1)
	/tmp/dragonfly/build/src/github.com/alibaba/Dragonfly/vendor/github.com/sirupsen/logrus/logger.go:244 +0x6d
github.com/alibaba/Dragonfly/dfget/config.AssertContext.func1(0xc420166000)
	/tmp/dragonfly/build/src/github.com/alibaba/Dragonfly/dfget/config/config.go:281 +0x87
panic(0x728f40, 0xc42007ff30)
	/usr/local/go/src/runtime/panic.go:502 +0x229
github.com/alibaba/Dragonfly/dfget/util.PanicIfError(0x7e17e0, 0xc42007ff10, 0x79e00b, 0xb)
	/tmp/dragonfly/build/src/github.com/alibaba/Dragonfly/dfget/util/util.go:77 +0x105
github.com/alibaba/Dragonfly/dfget/config.AssertContext(0xc420166000)
	/tmp/dragonfly/build/src/github.com/alibaba/Dragonfly/dfget/config/config.go:285 +0x121
github.com/alibaba/Dragonfly/dfget/cli.initialize()
	/tmp/dragonfly/build/src/github.com/alibaba/Dragonfly/dfget/cli/cli.go:50 +0x3b
github.com/alibaba/Dragonfly/dfget/cli.Run()
	/tmp/dragonfly/build/src/github.com/alibaba/Dragonfly/dfget/cli/cli.go:37 +0x26
main.main()
	/tmp/dragonfly/build/src/github.com/alibaba/Dragonfly/cmd/dfget/main.go:28 +0x20
```
When I use a invalid URL and get panic, it's not make sense for end user.

WDYT ?

--- a/cmd/dfget/app/root.go
+++ b/cmd/dfget/app/root.go
@@ -13,6 +13,7 @@ import (
 	"github.com/dragonflyoss/Dragonfly/dfget/errors"
 	"github.com/dragonflyoss/Dragonfly/dfget/util"
 
+	errHandler "github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 	"github.com/spf13/cobra"
 )
@@ -41,22 +42,34 @@ func init() {
 
 // runDfget do some init operations and start to download.
 func runDfget() error {
-	// initialize logger and get properties
-	initLog()
+	// initialize logger
+	if err := initClientLog(); err != nil {
+		util.Printer.Println(fmt.Sprintf("init log error: %v", err))
+		return err
+	}
+
+	// get config from property files
 	initProperties()
 
-	// check the legitimacy of parameters
+	if err := transParams(); err != nil {
+		util.Printer.Println(err.Error())
+		return err
+	}
+
 	checkParameters()
 	cfg.ClientLogger.Infof("get cmd params:%q", os.Args)
 
-	config.AssertConfig(cfg)
+	if err := config.AssertConfig(cfg); err != nil {
+		util.Printer.Println(fmt.Sprintf("assert context error: %v", err))
+		return err
+	}
 	cfg.ClientLogger.Infof("get init config:%v", cfg)
 
 	// enter the core process
 	err := core.Start(cfg)
 	util.Printer.Println(resultMsg(cfg, time.Now(), err))
 	if err != nil {
-		return err
+		os.Exit(err.Code)
 	}
 	return nil
 }
@@ -68,6 +81,7 @@ func checkParameters() {
 	}
 }
 
+// load config from property files.
 func initProperties() {
 	properties := config.NewProperties()
 	for _, v := range cfg.ConfigFiles {
@@ -94,17 +108,27 @@ func initProperties() {
 	if cfg.ClientQueueSize == 0 {
 		cfg.ClientQueueSize = properties.ClientQueueSize
 	}
+}
 
+// transParams trans the user-friendly parameter formats
+// to the format corresponding to the `Config` struct.
+func transParams() error {
 	cfg.Filter = transFilter(filter)
 
 	var err error
-	cfg.LocalLimit, err = transLimit(localLimit)
-	util.PanicIfError(err, "convert locallimit error")
-	cfg.TotalLimit, err = transLimit(totalLimit)
-	util.PanicIfError(err, "convert totallimit error")
+	if cfg.LocalLimit, err = transLimit(localLimit); err != nil {
+		return errHandler.Wrapf(errors.ErrConvertFailed, "locallimit: %v", err)
+	}
+
+	if cfg.TotalLimit, err = transLimit(totalLimit); err != nil {
+		return errHandler.Wrapf(errors.ErrConvertFailed, "totallimit: %v", err)
+	}
+
+	return nil
 }
 
-func initLog() {
+// initClientLog init client logger.
+func initClientLog() error {
 	var (
 		logPath  = path.Join(cfg.WorkHome, "logs")
 		logLevel = "info"
@@ -112,16 +136,27 @@ func initLog() {
 	if cfg.Verbose {
 		logLevel = "debug"
 	}
-	cfg.ClientLogger = util.CreateLogger(logPath, "dfclient.log", logLevel, cfg.Sign)
+
+	var err error
+	if cfg.ClientLogger, err = util.CreateLogger(logPath, "dfclient.log", logLevel, cfg.Sign); err != nil {
+		return err
+	}
+
 	if cfg.Console {
 		util.AddConsoleLog(cfg.ClientLogger)
 	}
-	if cfg.Pattern == config.PatternP2P {
-		cfg.ServerLogger = util.CreateLogger(logPath, "dfserver.log", logLevel, cfg.Sign)
-	}
+	return nil
 }
 
 func initFlags() {
+	persistFlagSet := rootCmd.PersistentFlags()
+
+	// pass to server
+	persistFlagSet.DurationVar(&cfg.RV.DataExpireTime, "expiretime", config.DataExpireTime,
+		"server will delete cached files if these files doesn't be modification within this duration")
+	persistFlagSet.DurationVar(&cfg.RV.ServerAliveTime, "alivetime", config.ServerAliveTime,
+		"server will stop if there is no uploading task within this duration")
+
 	flagSet := rootCmd.Flags()
 
 	// url & output
@@ -133,7 +168,7 @@ func initFlags() {
 	// localLimit & totalLimit & timeout
 	flagSet.StringVarP(&localLimit, "locallimit", "s", "",
 		"rate limit about a single download task, its format is 20M/m/K/k")
-	flagSet.StringVarP(&totalLimit, "totallimit", "", "",
+	flagSet.StringVar(&totalLimit, "totallimit", "",
 		"rate limit about the whole host, its format is 20M/m/K/k")
 	flagSet.IntVarP(&cfg.Timeout, "timeout", "e", 0,
 		"download timeout(second)")
@@ -166,13 +201,6 @@ func initFlags() {
 		"not back source when p2p fail")
 	flagSet.BoolVar(&cfg.DFDaemon, "dfdaemon", false,
 		"caller is from dfdaemon")
-
-	// pass to server
-	rootCmd.PersistentFlags().DurationVar(&cfg.RV.DataExpireTime, "expiretime", config.DataExpireTime,
-		"server will delete cached files if these files doesn't be modification within this duration")
-	rootCmd.PersistentFlags().DurationVar(&cfg.RV.ServerAliveTime, "alivetime", config.ServerAliveTime,
-		"server will stop if there is no uploading task within this duration")
-
 	// others
 	flagSet.BoolVarP(&cfg.ShowBar, "showbar", "b", false,
 		"show progress bar, it's conflict with '--console'")

--- a/cmd/dfget/app/server.go
+++ b/cmd/dfget/app/server.go
@@ -21,8 +21,10 @@ import (
 	"os"
 	"path/filepath"
 
+	"github.com/dragonflyoss/Dragonfly/dfget/config"
 	"github.com/dragonflyoss/Dragonfly/dfget/core/uploader"
 	"github.com/dragonflyoss/Dragonfly/dfget/util"
+
 	"github.com/spf13/cobra"
 )
 
@@ -30,12 +32,19 @@ var serverCmd = &cobra.Command{
 	Use:   "server",
 	Short: "Launch a peer server for uploading files.",
 	Run: func(cmd *cobra.Command, args []string) {
-		cfg.ServerLogger = util.CreateLogger(filepath.Join(cfg.WorkHome, "logs"),
-			"dfserver.log", "INFO", cfg.Sign)
+		// init server logger.
+		serverLogger, err := util.CreateLogger(filepath.Join(cfg.WorkHome, "logs"), "dfserver.log", "INFO", cfg.Sign)
+		if err != nil {
+			fmt.Println(err.Error())
+			os.Exit(config.CodeLaunchServerError)
+		}
+		cfg.ServerLogger = serverLogger
+
+		// launch a peer server as a uploader server
 		port, err := uploader.LaunchPeerServer(cfg)
 		if err != nil {
 			fmt.Println(err.Error())
-			os.Exit(11)
+			os.Exit(config.CodeLaunchServerError)
 		}
 		fmt.Println(port)
 		uploader.WaitForShutdown()
@@ -48,14 +57,16 @@ func init() {
 }
 
 func initServerFlags() {
-	serverCmd.PersistentFlags().StringVar(&cfg.RV.SystemDataDir, "data", cfg.RV.SystemDataDir,
+	flagSet := serverCmd.Flags()
+
+	flagSet.StringVar(&cfg.RV.SystemDataDir, "data", cfg.RV.SystemDataDir,
 		"the directory which stores temporary files for p2p uploading")
-	serverCmd.PersistentFlags().StringVar(&cfg.WorkHome, "home", cfg.WorkHome,
+	flagSet.StringVar(&cfg.WorkHome, "home", cfg.WorkHome,
 		"the work home of dfget server")
-	serverCmd.PersistentFlags().StringVar(&cfg.RV.LocalIP, "ip", "",
+	flagSet.StringVar(&cfg.RV.LocalIP, "ip", "",
 		"the ip that server will listen on")
-	serverCmd.PersistentFlags().IntVar(&cfg.RV.PeerPort, "port", 0,
+	flagSet.IntVar(&cfg.RV.PeerPort, "port", 0,
 		"the port that server will listen on")
-	serverCmd.PersistentFlags().StringVar(&cfg.RV.MetaPath, "meta", cfg.RV.MetaPath,
+	flagSet.StringVar(&cfg.RV.MetaPath, "meta", cfg.RV.MetaPath,
 		"meta file path")
 }

--- a/dfget/config/config.go
+++ b/dfget/config/config.go
@@ -30,7 +30,10 @@ import (
 	"syscall"
 	"time"
 
+	errType "github.com/dragonflyoss/Dragonfly/dfget/errors"
 	"github.com/dragonflyoss/Dragonfly/dfget/util"
+
+	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 	"gopkg.in/gcfg.v1"
 	"gopkg.in/warnings.v0"
@@ -240,35 +243,46 @@ func NewConfig() *Config {
 	cfg.Sign = fmt.Sprintf("%d-%.3f",
 		os.Getpid(), float64(time.Now().UnixNano())/float64(time.Second))
 
-	if currentUser, err := user.Current(); err == nil {
-		cfg.User = currentUser.Username
-		cfg.WorkHome = path.Join(currentUser.HomeDir, ".small-dragonfly")
-		cfg.RV.MetaPath = path.Join(cfg.WorkHome, "meta", "host.meta")
-		cfg.RV.SystemDataDir = path.Join(cfg.WorkHome, "data")
-		cfg.RV.FileLength = -1
-	} else {
-		panic(fmt.Errorf("get user error: %s", err))
+	// TODO: Use parameters instead of currentUser.HomeDir.
+	currentUser, err := user.Current()
+	if err != nil {
+		util.Printer.Println(fmt.Sprintf("get user error: %s", err))
+		os.Exit(CodeGetUserError)
 	}
+
+	cfg.User = currentUser.Username
+	cfg.WorkHome = path.Join(currentUser.HomeDir, ".small-dragonfly")
+	cfg.RV.MetaPath = path.Join(cfg.WorkHome, "meta", "host.meta")
+	cfg.RV.SystemDataDir = path.Join(cfg.WorkHome, "data")
+	cfg.RV.FileLength = -1
 	cfg.ConfigFiles = []string{DefaultYamlConfigFile, DefaultIniConfigFile}
 	return cfg
 }
 
-// AssertConfig checks the config and panic if any error happens.
-func AssertConfig(cfg *Config) {
-	util.PanicIfNil(cfg, "runtime config is not initialized")
-	util.PanicIfNil(cfg.ClientLogger, "client log is not initialized")
-	if cfg.Pattern == "p2p" {
-		util.PanicIfNil(cfg.ServerLogger, "server log is not initialized")
-	}
-
+// AssertConfig checks the config and return errors.
+func AssertConfig(cfg *Config) (err error) {
 	defer func() {
-		if err := recover(); err != nil {
-			cfg.ClientLogger.Panic(err)
+		if err != nil {
+			cfg.ClientLogger.Errorf("assert context error: %v", err)
 		}
 	}()
 
-	util.PanicIfError(checkURL(cfg), "invalid url")
-	util.PanicIfError(checkOutput(cfg), "invalid output")
+	if util.IsNil(cfg) {
+		return errors.Wrap(errType.ErrNotInitialized, "runtime config")
+	}
+
+	if util.IsNil(cfg.ClientLogger) {
+		return errors.Wrap(errType.ErrNotInitialized, "client log")
+	}
+
+	if err := checkURL(cfg); err != nil {
+		return errors.Wrapf(errType.ErrInvalidValue, "url: %v", err)
+	}
+
+	if err := checkOutput(cfg); err != nil {
+		return err
+	}
+	return nil
 }
 
 func checkURL(cfg *Config) error {

--- a/dfget/config/config_test.go
+++ b/dfget/config/config_test.go
@@ -29,12 +29,14 @@ import (
 	"testing"
 	"time"
 
+	"github.com/dragonflyoss/Dragonfly/dfget/errors"
 	"github.com/dragonflyoss/Dragonfly/dfget/util"
+
 	"github.com/go-check/check"
 	"github.com/sirupsen/logrus"
 )
 
-var Cfg = NewConfig()
+var cfg = NewConfig()
 
 func Test(t *testing.T) {
 	check.TestingT(t)
@@ -52,35 +54,35 @@ func (suite *ConfigSuite) SetUpTest(c *check.C) {
 
 func (suite *ConfigSuite) TestConfig_String(c *check.C) {
 	expected := "{\"url\":\"\",\"output\":\"\""
-	c.Assert(strings.Contains(Cfg.String(), expected), check.Equals, true)
-	Cfg.LocalLimit = 20971520
-	Cfg.Pattern = "p2p"
-	Cfg.Version = true
+	c.Assert(strings.Contains(cfg.String(), expected), check.Equals, true)
+	cfg.LocalLimit = 20971520
+	cfg.Pattern = "p2p"
+	cfg.Version = true
 	expected = "\"url\":\"\",\"output\":\"\",\"localLimit\":20971520," +
 		"\"pattern\":\"p2p\",\"version\":true"
-	c.Assert(strings.Contains(Cfg.String(), expected), check.Equals, true)
+	c.Assert(strings.Contains(cfg.String(), expected), check.Equals, true)
 }
 
 func (suite *ConfigSuite) TestNewConfig(c *check.C) {
 	before := time.Now()
 	time.Sleep(time.Millisecond)
-	Cfg = NewConfig()
+	cfg := NewConfig()
 	time.Sleep(time.Millisecond)
 	after := time.Now()
 
-	c.Assert(Cfg.StartTime.After(before), check.Equals, true)
-	c.Assert(Cfg.StartTime.Before(after), check.Equals, true)
+	c.Assert(cfg.StartTime.After(before), check.Equals, true)
+	c.Assert(cfg.StartTime.Before(after), check.Equals, true)
 
 	beforeSign := fmt.Sprintf("%d-%.3f",
 		os.Getpid(), float64(before.UnixNano())/float64(time.Second))
 	afterSign := fmt.Sprintf("%d-%.3f",
 		os.Getpid(), float64(after.UnixNano())/float64(time.Second))
-	c.Assert(beforeSign < Cfg.Sign, check.Equals, true)
-	c.Assert(afterSign > Cfg.Sign, check.Equals, true)
+	c.Assert(beforeSign < cfg.Sign, check.Equals, true)
+	c.Assert(afterSign > cfg.Sign, check.Equals, true)
 
 	if curUser, err := user.Current(); err != nil {
-		c.Assert(Cfg.User, check.Equals, curUser.Username)
-		c.Assert(Cfg.WorkHome, check.Equals, path.Join(curUser.HomeDir, ".small-dragonfly"))
+		c.Assert(cfg.User, check.Equals, curUser.Username)
+		c.Assert(cfg.WorkHome, check.Equals, path.Join(curUser.HomeDir, ".small-dragonfly"))
 	}
 }
 
@@ -92,45 +94,32 @@ func (suite *ConfigSuite) TestAssertConfig(c *check.C) {
 	clog.Out = buf
 
 	var cases = []struct {
-		clog     *logrus.Logger
-		slog     *logrus.Logger
-		url      string
-		output   string
-		expected string
+		clog      *logrus.Logger
+		slog      *logrus.Logger
+		url       string
+		output    string
+		checkFunc func(err error) bool
 	}{
-		{expected: "client log"},
-		{clog: clog, expected: "server log"},
-		{clog: clog, slog: clog, expected: "invalid url"},
-		{clog: clog, slog: clog, url: "http://a.b", expected: ""},
-		{clog: clog, slog: clog, url: "http://a.b", output: "/root", expected: "invalid output"},
+		{checkFunc: errors.IsNotInitialized},
+		{clog: clog, checkFunc: errors.IsNotInitialized},
+		{clog: clog, slog: clog, checkFunc: errors.IsInvalidValue},
+		{clog: clog, slog: clog, url: "http://a.b", checkFunc: errors.IsNilError},
+		{clog: clog, slog: clog, url: "http://a.b", output: "/root", checkFunc: errors.IsInvalidValue},
 	}
 
-	var f = func() (msg string) {
-		defer func() {
-			if r := recover(); r != nil {
-				switch r := r.(type) {
-				case error:
-					msg = r.Error()
-				case *logrus.Entry:
-					msg = r.Message
-					buf.Reset()
-				default:
-					msg = fmt.Sprintf("%v", r)
-				}
-			}
-		}()
-		AssertConfig(Cfg)
-		return ""
+	var f = func() (err error) {
+		return AssertConfig(cfg)
 	}
 
 	for _, v := range cases {
-		Cfg.ClientLogger = v.clog
-		Cfg.ServerLogger = v.slog
-		Cfg.URL = v.url
-		Cfg.Output = v.output
+		cfg.ClientLogger = v.clog
+		cfg.ServerLogger = v.slog
+		cfg.URL = v.url
+		cfg.Output = v.output
 		actual := f()
-		c.Assert(strings.HasPrefix(actual, v.expected), check.Equals, true,
-			check.Commentf("actual:[%s] expected:[%s]", actual, v.expected))
+		expected := v.checkFunc(actual)
+		c.Assert(expected, check.Equals, true,
+			check.Commentf("actual:[%s] expected:[%s]", actual, expected))
 	}
 }
 
@@ -156,11 +145,11 @@ func (suite *ConfigSuite) TestCheckURL(c *check.C) {
 			"q=is%3Aissue+is%3Aclosed": true,
 	}
 
-	c.Assert(checkURL(Cfg), check.NotNil)
+	c.Assert(checkURL(cfg), check.NotNil)
 	for k, v := range cases {
 		for _, scheme := range []string{"http", "https", "HTTP", "HTTPS"} {
-			Cfg.URL = fmt.Sprintf("%s://%s", scheme, k)
-			actual := fmt.Sprintf("%s:%v", k, checkURL(Cfg))
+			cfg.URL = fmt.Sprintf("%s://%s", scheme, k)
+			actual := fmt.Sprintf("%s:%v", k, checkURL(cfg))
 			expected := fmt.Sprintf("%s:%s://%s", k, scheme, k)
 			if v {
 				expected = fmt.Sprintf("%s:<nil>", k)
@@ -190,17 +179,17 @@ func (suite *ConfigSuite) TestCheckOutput(c *check.C) {
 		{"", "/tmp/a/b/c/d/e/zj.test", "/tmp/a/b/c/d/e/zj.test"},
 	}
 
-	if Cfg.User != "root" {
+	if cfg.User != "root" {
 		cases = append(cases, tester{url: "", output: "/root/zj.test", expected: ""})
 	}
 	for _, v := range cases {
-		Cfg.URL = v.url
-		Cfg.Output = v.output
+		cfg.URL = v.url
+		cfg.Output = v.output
 		if util.IsEmptyStr(v.expected) {
-			c.Assert(checkOutput(Cfg), check.NotNil, check.Commentf("%v", v))
+			c.Assert(checkOutput(cfg), check.NotNil, check.Commentf("%v", v))
 		} else {
-			c.Assert(checkOutput(Cfg), check.IsNil, check.Commentf("%v", v))
-			c.Assert(Cfg.Output, check.Equals, v.expected, check.Commentf("%v", v))
+			c.Assert(checkOutput(cfg), check.IsNil, check.Commentf("%v", v))
+			c.Assert(cfg.Output, check.Equals, v.expected, check.Commentf("%v", v))
 		}
 	}
 }

--- a/dfget/config/constants.go
+++ b/dfget/config/constants.go
@@ -109,3 +109,21 @@ const (
 	DataExpireTime  = 3 * time.Minute
 	ServerAliveTime = 5 * time.Minute
 )
+
+/* errors code */
+const (
+	// CodeLaunchServerError represents failed to launch a peer server.
+	CodeLaunchServerError = 1100 + iota
+
+	// CodePrepareError represents failed to prepare before downloading.
+	CodePrepareError
+
+	// CodeGetUserError represents failed to get current user.
+	CodeGetUserError
+
+	// CodeRegisterError represents failed to register to supernode.
+	CodeRegisterError
+
+	// CodeDownloadError represents failed to download file.
+	CodeDownloadError
+)

--- a/dfget/core/core.go
+++ b/dfget/core/core.go
@@ -57,27 +57,22 @@ func Start(cfg *config.Config) *errors.DFGetError {
 		cfg.StartTime.Format(config.DefaultTimestampFormat), cfg.URL))
 
 	if err = prepare(cfg); err != nil {
-		return errors.New(1100, err.Error())
+		return errors.New(config.CodePrepareError, err.Error())
 	}
 
 	if result, err = registerToSuperNode(cfg, register); err != nil {
-		return errors.New(1200, err.Error())
+		return errors.New(config.CodeRegisterError, err.Error())
 	}
 
 	if err = downloadFile(cfg, supernodeAPI, register, result); err != nil {
-		return errors.New(1300, err.Error())
+		return errors.New(config.CodeDownloadError, err.Error())
 	}
 
 	return nil
 }
 
+// prepare the RV-related information and create the corresponding files.
 func prepare(cfg *config.Config) (err error) {
-	defer func() {
-		if r := recover(); r != nil {
-			err = r.(error)
-		}
-	}()
-
 	util.Printer.Printf("dfget version:%s", version.DFGetVersion)
 	util.Printer.Printf("workspace:%s sign:%s", cfg.WorkHome, cfg.Sign)
 	cfg.ClientLogger.Infof("target file path:%s", cfg.Output)
@@ -86,13 +81,23 @@ func prepare(cfg *config.Config) (err error) {
 
 	rv.RealTarget = cfg.Output
 	rv.TargetDir = path.Dir(rv.RealTarget)
-	panicIf(util.CreateDirectory(rv.TargetDir))
-	cfg.RV.TempTarget, err = createTempTargetFile(rv.TargetDir, cfg.Sign)
-	panicIf(err)
+	if err = util.CreateDirectory(rv.TargetDir); err != nil {
+		return err
+	}
 
-	panicIf(util.CreateDirectory(path.Dir(rv.MetaPath)))
-	panicIf(util.CreateDirectory(cfg.WorkHome))
-	panicIf(util.CreateDirectory(rv.SystemDataDir))
+	if cfg.RV.TempTarget, err = createTempTargetFile(rv.TargetDir, cfg.Sign); err != nil {
+		return err
+	}
+
+	if err = util.CreateDirectory(path.Dir(rv.MetaPath)); err != nil {
+		return err
+	}
+	if err = util.CreateDirectory(cfg.WorkHome); err != nil {
+		return err
+	}
+	if err = util.CreateDirectory(rv.SystemDataDir); err != nil {
+		return err
+	}
 	rv.DataDir = cfg.RV.SystemDataDir
 
 	cfg.Node = adjustSupernodeList(cfg.Node)
@@ -287,10 +292,4 @@ func calculateTimeout(fileLength int64, defaultTimeoutSecond int) time.Duration 
 		timeout = int(fileLength/(64*1024) + 10)
 	}
 	return time.Duration(timeout) * time.Second
-}
-
-func panicIf(err error) {
-	if err != nil {
-		panic(err)
-	}
 }

--- a/dfget/core/core_test.go
+++ b/dfget/core/core_test.go
@@ -85,16 +85,16 @@ func (s *CoreTestSuite) TestRegisterToSupernode(c *check.C) {
 		}
 	}
 
-	f(config.BackSourceReasonNodeEmpty, true, nil)
+	f(config.BackSourceReasonNodeEmpty, false, nil)
 
 	cfg.Pattern = config.PatternSource
-	f(config.BackSourceReasonUserSpecified, true, nil)
+	f(config.BackSourceReasonUserSpecified, false, nil)
 
 	cfg.Pattern = config.PatternP2P
 
 	cfg.Node = []string{"x"}
 	cfg.URL = "http://x.com"
-	f(config.BackSourceReasonRegisterFail, true, nil)
+	f(config.BackSourceReasonRegisterFail, false, nil)
 
 	cfg.Node = []string{"x"}
 	cfg.URL = "http://taobao.com"

--- a/dfget/core/uploader/uploader.go
+++ b/dfget/core/uploader/uploader.go
@@ -53,6 +53,8 @@ var (
 	aliveQueue  = util.NewQueue(0)
 )
 
+// TODO: Move this part out of the uploader
+
 // StartPeerServerProcess starts an independent peer server process for uploading downloaded files
 // if it doesn't exist.
 // This function is invoked when dfget starts to download files in p2p pattern.
@@ -119,15 +121,15 @@ func checkPeerServerExist(cfg *config.Config, port int) int {
 
 	// check the peer server whether is available
 	result, err := checkServer(cfg.RV.LocalIP, port, cfg.RV.TargetDir, taskFileName, 0)
-	cfg.ServerLogger.Infof("local http result:%s err:%v, port:%d path:%s",
+	cfg.ClientLogger.Infof("local http result:%s err:%v, port:%d path:%s",
 		result, err, port, config.LocalHTTPPathCheck)
 
 	if err == nil {
 		if result == taskFileName {
-			cfg.ServerLogger.Infof("use peer server on port:%d", port)
+			cfg.ClientLogger.Infof("use peer server on port:%d", port)
 			return port
 		}
-		cfg.ServerLogger.Warnf("not found process on port:%d, version:%s", port, version.DFGetVersion)
+		cfg.ClientLogger.Warnf("not found process on port:%d, version:%s", port, version.DFGetVersion)
 	}
 	return 0
 }

--- a/dfget/errors/errors.go
+++ b/dfget/errors/errors.go
@@ -19,6 +19,25 @@ package errors
 
 import (
 	"fmt"
+
+	errHandler "github.com/pkg/errors"
+)
+
+var (
+	// ErrInvalidValue represents the value is invalid.
+	ErrInvalidValue = &DFGetError{codeInvalidValue, "invalid value"}
+
+	// ErrNotInitialized represents the object is not initialized.
+	ErrNotInitialized = &DFGetError{codeNotInitialized, "not initialized"}
+
+	// ErrConvertFailed represents failed to convert.
+	ErrConvertFailed = &DFGetError{codeConvertFailed, "convert failed"}
+)
+
+const (
+	codeInvalidValue = iota
+	codeNotInitialized
+	codeConvertFailed
 )
 
 // New function creates a DFGetError.
@@ -46,4 +65,35 @@ type DFGetError struct {
 
 func (e *DFGetError) Error() string {
 	return fmt.Sprintf("{\"Code\":%d,\"Msg\":\"%s\"}", e.Code, e.Msg)
+}
+
+// IsNilError check the error is nil or not.
+func IsNilError(err error) bool {
+	if err == nil {
+		return true
+	}
+	return false
+}
+
+// IsInvalidValue check the error is the value is invalid or not.
+func IsInvalidValue(err error) bool {
+	return checkError(err, codeInvalidValue)
+}
+
+// IsNotInitialized check the error is the object is not initialized or not.
+func IsNotInitialized(err error) bool {
+	return checkError(err, codeNotInitialized)
+}
+
+// IsConvertFailed check the error is a conversion error or not.
+func IsConvertFailed(err error) bool {
+	return checkError(err, codeConvertFailed)
+}
+
+func checkError(err error, code int) bool {
+	errCause := errHandler.Cause(err)
+	if errTemp, ok := errCause.(*DFGetError); ok && errTemp.Code == code {
+		return true
+	}
+	return false
 }

--- a/dfget/util/log.go
+++ b/dfget/util/log.go
@@ -31,27 +31,32 @@ import (
 const DefaultLogTimeFormat = "2006-01-02 15:04:05.000"
 
 // CreateLogger creates a logger.
-func CreateLogger(logPath string, logName string, logLevel string, sign string) *log.Logger {
-	var (
-		logger      *log.Logger
-		logFilePath = path.Join(logPath, logName)
-		level, err  = log.ParseLevel(logLevel)
-	)
+func CreateLogger(logPath string, logName string, logLevel string, sign string) (*log.Logger, error) {
+	// parse log level
+	level, err := log.ParseLevel(logLevel)
 	if err != nil {
+		// TODO: print a warning.
 		level = log.InfoLevel
 	}
-	if err = os.MkdirAll(filepath.Dir(logFilePath), 0755); err == nil {
-		var logFile *os.File
-		if logFile, err = os.OpenFile(logFilePath, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0644); err == nil {
-			logFile.Seek(0, 2)
-			logger = log.New()
-			logger.Out = logFile
-			logger.Formatter = &DragonflyFormatter{TimestampFormat: DefaultLogTimeFormat, Sign: sign}
-			logger.Level = level
-			return logger
-		}
+
+	// create log file path
+	logFilePath := path.Join(logPath, logName)
+	if err := os.MkdirAll(filepath.Dir(logFilePath), 0755); err != nil {
+		return nil, err
 	}
-	panic(err)
+
+	// open log file
+	logFile, err := os.OpenFile(logFilePath, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0644)
+	if err != nil {
+		return nil, err
+	}
+
+	logFile.Seek(0, 2)
+	logger := log.New()
+	logger.Out = logFile
+	logger.Formatter = &DragonflyFormatter{TimestampFormat: DefaultLogTimeFormat, Sign: sign}
+	logger.Level = level
+	return logger, nil
 }
 
 // AddConsoleLog will add a ConsoleLog into logger's hooks.

--- a/dfget/util/log_test.go
+++ b/dfget/util/log_test.go
@@ -34,6 +34,8 @@ func (suite *DFGetUtilSuite) TestCreateLogger(c *check.C) {
 	logger, tmpFile, r, err := tempFileAndLogger("debug", "x")
 	defer cleanTempFile(tmpFile, err)
 
+	c.Assert(err, check.IsNil)
+
 	var checkLogs = func(level logrus.Level, msg string) {
 		line, _, _ := r.ReadLine()
 		tmpStr := strings.Split(strings.Trim(string(line), "\n"), " ")
@@ -43,38 +45,24 @@ func (suite *DFGetUtilSuite) TestCreateLogger(c *check.C) {
 		c.Assert(strings.Join(tmpStr[5:], " "), check.Equals, msg)
 	}
 
-	var testPanic = func() {
-		msg := "panic test"
-		defer func() {
-			if e := recover(); e != nil {
-				s := e.(*logrus.Entry)
-				c.Assert(s.Level, check.Equals, logrus.PanicLevel)
-				c.Assert(s.Message, check.Equals, msg)
-			}
-		}()
-		logger.Panic(msg)
-	}
-
 	logger.Debug("test")
 	checkLogs(logrus.DebugLevel, "test")
 	logger.Info([]int{1, 2, 3})
 	checkLogs(logrus.InfoLevel, "[1 2 3]")
 	logger.Warn("test")
 	checkLogs(logrus.WarnLevel, "test")
-
-	testPanic()
 }
 
 func (suite *DFGetUtilSuite) TestCreateLogger_differentLevel(c *check.C) {
 	var (
 		tmpPath     = "/tmp"
 		tmpFileName = fmt.Sprintf("dfget_test_%d.log", rand.Int63())
-		logger      *logrus.Logger
 	)
 	defer os.Remove(path.Join(tmpPath, tmpFileName))
 
 	var testLevel = func(level string, expected logrus.Level) {
-		logger = CreateLogger(tmpPath, tmpFileName, level, "")
+		logger, err := CreateLogger(tmpPath, tmpFileName, level, "")
+		c.Assert(err, check.IsNil)
 		c.Assert(logger.Level, check.Equals, expected)
 	}
 
@@ -91,6 +79,8 @@ func (suite *DFGetUtilSuite) TestCreateLogger_differentLevel(c *check.C) {
 func (suite *DFGetUtilSuite) TestAddConsoleLog(c *check.C) {
 	logger, tmpFile, r, err := tempFileAndLogger("debug", "")
 	defer cleanTempFile(tmpFile, err)
+
+	c.Assert(err, check.IsNil)
 
 	// remove timestamp from logs to avoid testing failure
 	logger.Formatter = &logrus.TextFormatter{DisableTimestamp: true}
@@ -123,9 +113,13 @@ func (suite *DFGetUtilSuite) TestAddConsoleLog(c *check.C) {
 func tempFileAndLogger(level string, sign string) (*logrus.Logger, *os.File, *bufio.Reader, error) {
 	tmpPath := "/tmp"
 	tmpFile, err := ioutil.TempFile(tmpPath, "dfget_test")
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
 	tmpFileName := strings.TrimLeft(tmpFile.Name(), "/tmp/")
 	r := bufio.NewReader(tmpFile)
-	logger := CreateLogger(tmpPath, tmpFileName, level, sign)
+	logger, err := CreateLogger(tmpPath, tmpFileName, level, sign)
 	return logger, tmpFile, r, err
 }
 

--- a/dfget/util/util.go
+++ b/dfget/util/util.go
@@ -20,7 +20,6 @@ package util
 
 import (
 	"encoding/json"
-	"fmt"
 	"os"
 	"reflect"
 )
@@ -62,20 +61,6 @@ func IsNil(value interface{}) (result bool) {
 		}
 	}
 	return
-}
-
-// PanicIfNil panic if the obj is nil.
-func PanicIfNil(obj interface{}, msg string) {
-	if IsNil(obj) {
-		panic(fmt.Errorf(msg))
-	}
-}
-
-// PanicIfError panic if an error happened.
-func PanicIfError(err error, msg string) {
-	if err != nil {
-		panic(fmt.Errorf("%s: %v", msg, err))
-	}
 }
 
 // JSONString returns json string of the v.

--- a/dfget/util/util_test.go
+++ b/dfget/util/util_test.go
@@ -17,7 +17,6 @@
 package util
 
 import (
-	"fmt"
 	"math/rand"
 	"testing"
 	"time"
@@ -64,39 +63,6 @@ func (suite *DFGetUtilSuite) TestIsNil(c *check.C) {
 
 	var temp *DFGetUtilSuite
 	c.Assert(IsNil(temp), check.Equals, true)
-}
-
-func (suite *DFGetUtilSuite) TestPanicIfNil(c *check.C) {
-	var f = func(v interface{}, msg string) (retMsg string) {
-		defer func() {
-			if r := recover(); r != nil {
-				retMsg = r.(error).Error()
-			}
-		}()
-
-		PanicIfNil(v, msg)
-		return ""
-	}
-
-	c.Assert(f(1, "int"), check.Equals, "")
-	c.Assert(f("", "string"), check.Equals, "")
-	c.Assert(f(nil, "nil"), check.Equals, "nil")
-	c.Assert(f(suite, "*DFGetUtilSuite"), check.Equals, "")
-}
-
-func (suite *DFGetUtilSuite) TestPanicIfError(c *check.C) {
-	var f = func(v error, msg string) (retMsg string) {
-		defer func() {
-			if r := recover(); r != nil {
-				retMsg = r.(error).Error()
-			}
-		}()
-
-		PanicIfError(v, msg)
-		return ""
-	}
-	c.Assert(f(nil, ""), check.Equals, "")
-	c.Assert(f(fmt.Errorf("test"), "error"), check.Equals, "error: test")
 }
 
 func (suite *DFGetUtilSuite) TestJsonString(c *check.C) {

